### PR TITLE
AirfRANS 4L/256d: lr=1e-3 + gc=1.5 compound effect

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
At 3L/192d, lr=1e-3 was stable with gc=1.5 (val 0.008582). At 4L/256d with WD=1e-2 providing L2 regularization, the combination of higher LR + higher grad clip may allow faster escape from local minima while the larger architecture provides more capacity to exploit those escapes. This tests whether the golden config's lr=7e-4 is overly conservative.

## Instructions
Run the following command exactly:

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset airfrans --airfrans_task full --optimizer adamw \
  --lr 1e-3 --cosine-t-max 5 --grad-clip 1.5 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --wandb_group airfrans-4L256d-compound
```

Key changes from golden config: `--lr 1e-3` (was 7e-4) and `--grad-clip 1.5` (was 1.0). All other settings match the proven 4L/256d golden config.

## Baseline
Current best metrics from BASELINE.md (PR #2727, W&B run: ruurxdqs):
- val_primary/surface_mse: **0.007264** (primary metric, lower is better)
- External target: 0.0043 (~1.7x gap remaining)
- Architecture: 4L/256d, 4 heads, AdamW lr=7e-4, gc=1.0, WD=1e-2, T_max=5, no-EMA, enable-fourier

Reproduce baseline:
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset airfrans --airfrans_task full --optimizer adamw \
  --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT armin:

## Results

### Experiment: AirfRANS 4L/256d lr=1e-3 + gc=1.5

| Metric | Baseline (PR #2727) | This Experiment | Delta |
|--------|-------------------|-----------------|-------|
| val_primary/surface_mse | **0.007264** | **0.1109** | +1427% (15.3x worse) |
| full_val/volume_mse | — | 0.1939 | — |
| full_val/surface_mse_p | — | 0.4430 | — |

**Best epoch**: 18 (of 31 run before early termination)

### Training Command
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw \
  --lr 1e-3 --cosine-t-max 5 --grad-clip 1.5 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --agent armin \
  --wandb-name "armin/airfrans-4L256d-lr1e3-gc15" \
  --wandb-group airfrans-4L256d-compound
```

- **W&B run**: `9m3auluz`
- **Peak memory**: ~8.5 GB VRAM (GPU 0)
- **Epochs completed**: 31 (terminated early due to catastrophic divergence)

### What happened — Catastrophic training instability

**This is a clear negative result.** The lr=1e-3 + gc=1.5 combination is far too aggressive for the 4L/256d architecture, causing catastrophic gradient explosion and training divergence.

**Key observations:**

1. **Gradient norm explosion**: Mean grad norm escalated catastrophically across training:
   - Epochs 1-10: 11→18 (moderate)
   - Epochs 11-20: 20→177 (escalating)
   - Epochs 20-25: 282→521→407 (severe)
   - Epochs 26-31: 448→574→**4158**→**1750** (catastrophic)

2. **Validation instability**: Val metric oscillated violently with no convergence:
   - Best: 0.1109 (epoch 18) — still 15.3x worse than baseline
   - Worst: 0.552 (epoch 30)
   - Peaks were GROWING, not dampening: 0.31→0.36→0.55→0.51

3. **100% grad clip saturation**: By epoch 19, all 360 batches per epoch were hitting the grad clip ceiling (360/360 events). The gc=1.5 threshold was overwhelmed — gradients were routinely 100-2700x above the clip value.

4. **Cosine trough analysis**: With T_max=5, troughs should occur every 5 epochs. The trough values (epochs 5, 10, 15, 20, 25, 30) were: 0.233→0.177→0.203→0.207→0.244→0.552. No improvement — the trough envelope was flat then diverging.

### Why it failed

The hypothesis assumed that WD=1e-2 would provide enough L2 regularization to stabilize lr=1e-3 at 4L/256d scale. This was incorrect:

- **At 3L/192d**: lr=1e-3 + gc=1.5 was stable (val 0.0086) because the smaller model has fewer parameters and shorter gradient paths, making it inherently more stable.
- **At 4L/256d**: The larger model (4 layers, 256 hidden dim, 4 heads) amplifies gradient instability through deeper backpropagation chains. The additional capacity that was supposed to help actually makes the training dynamics harder to control.
- **gc=1.5 was too loose**: At gc=1.0 (golden config), the tighter clip provides stronger implicit regularization. At gc=1.5, gradients can grow 50% larger per step, compounding through the deeper architecture.
- **The golden config's lr=7e-4 is NOT overly conservative** — it appears to be well-calibrated for 4L/256d. The successful PR #2719 result (val=0.00223) used lr=5e-4 with gc=1.0 on 3L/192d, suggesting that lower LR + tighter clip is the winning formula.

### Suggested follow-ups

1. **4L/256d with lr=5e-4 + gc=1.0**: Port the PR #2719 winning recipe to the larger architecture. If 4L/256d provides meaningful extra capacity, it should benefit from the same stabilized training dynamics.
2. **4L/256d lr=7e-4 + gc=0.5**: Try an even tighter grad clip than the golden config to see if more aggressive clipping helps the larger architecture converge deeper.
3. **Warmup for high LR**: If lr=1e-3 at 4L/256d is desired, a gradual LR warmup over the first 5-10 epochs might prevent the early gradient instability.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2776 because ADVISOR: Closing — lr=1e-3+gc=1.5 at 4L/256d diverged catastrophically (0.111, grad norms→4158). This combination is fundamentally incompatible at this scale..
